File: test/integration/service.cpp
//
// Created by Rakesh on 13/01/2022.
//

#include <catch2/catch.hpp>

#include "../../src/log/NanoLog.h"
#include "../../src/api/api.h"

namespace spt::encrypter::itest::connection
{
  std::string encrypted;
}

using namespace std::string_view_literals;
namespace api = spt::encrypter::api;

SCENARIO( "TCP service test", "tcp" )
{
  GIVEN( "Using the client API" )
  {
    const auto text = "Test string to be encrypted and decrypted"sv;

    WHEN( "Encrypting a string" )
    {
      auto enc = api::encrypt( text );
      REQUIRE_FALSE( enc.empty() );
      spt::encrypter::itest::connection::encrypted = enc;
    }

    AND_WHEN( "Decrypting a string" )
    {
      REQUIRE_FALSE( spt::encrypter::itest::connection::encrypted.empty() );
      const auto dec = api::decrypt( spt::encrypter::itest::connection::encrypted );
      REQUIRE( dec.size() == text.size() );
      REQUIRE( dec == text );
    }

    AND_WHEN( "Encrypting and decrypting within same context" )
    {
      const auto another = "0ENKfo4NkIAHVBoNUtUseShHK/6JQ11YGALUSA/RLcICLQaTyxDJcldEsxxM0h+AWdkXAXNY6xOhCk+TkobTayscjTPqu6bzKIxnD1hso8Hz/ZWsh4SlSxHLhHAZRg6VfBZlITexMDdMOks2eauDdND6ft5kT+7uJFpq5EfnuHE="sv;
      auto enc = api::encrypt( another );
      REQUIRE_FALSE( enc.empty() );
      const auto dec = api::decrypt( enc );
      REQUIRE( dec.size() == another.size() );
      REQUIRE( dec == another );
    }

    AND_WHEN( "Encrypting and decrypting in a loop" )
    {
      const auto another = R"(VvgRpEXRtq3AMUDdIO7GvFDfHgqZFC/n5zh72V7HEGn9rNIXTYbHGwGVjoSEev6h
dE6EezzqeeshfkDEO/Qqs1Csvzf+QrzqCaQOq40LuhpDdJym2SnuIGJoPDnuTu7/
Fu2DjHFSibtC/E1UokF6hwd5Wiq35BiV3Yzev99DTWs=)"sv;

      for ( auto i = 0; i < 10; ++i )
      {
        auto enc = api::encrypt( another );
        REQUIRE_FALSE( enc.empty() );
        const auto dec = api::decrypt( enc );
        REQUIRE( dec.size() == another.size() );
        REQUIRE( dec == another );
      }
    }
  }
}

File: client/python/features/steps/service.py
from base64 import b64encode, b64decode
from behave import given, then, step
from behave.api.async_step import async_run_until_complete
from behave.runner import Context
from hamcrest import assert_that, equal_to, not_

from client import Client as _Client


@given("The encrypter service running on {host} on port {port:d}")
def step_impl(context: Context, host: str, port: int):
    context.host = host
    context.port = port
    context.value = "0ENKfo4NkIAHVBoNUtUseShHK/6JQ11YGALUSA/RLcICLQaTyxDJcldEsxxM0h+AWdkXAXNY6xOhCk+TkobTayscjTPqu6bzKIxnD1hso8Hz/ZWsh4SlSxHLhHAZRg6VfBZlITexMDdMOks2eauDdND6ft5kT+7uJFpq5EfnuHE="


@then("Encrypt a string")
@async_run_until_complete
async def step_impl(context: Context):
    assert_that(hasattr(context, "host"), equal_to(True), "Encrypter service host not set")
    assert_that(hasattr(context, "port"), equal_to(True), "Encrypter service port not set")
    assert_that(hasattr(context, "value"), equal_to(True), "String to encrypt not set")

    async with _Client(host=context.host, port=context.port) as client:
        value = await client.encrypt(context.value)
        assert_that(len(value), not_(equal_to(0)), "Encrypted value of zero length")
        context.encrypted = value


@step("Decrypt the encrypted string")
@async_run_until_complete
async def step_impl(context: Context):
    assert_that(hasattr(context, "host"), equal_to(True), "Encrypter service host not set")
    assert_that(hasattr(context, "port"), equal_to(True), "Encrypter service port not set")
    assert_that(hasattr(context, "value"), equal_to(True), "String to encrypt not set")
    assert_that(hasattr(context, "encrypted"), equal_to(True), "Encrypted string not set")

    async with _Client(host=context.host, port=context.port) as client:
        value = await client.decrypt(context.encrypted)
        assert_that(value, equal_to(context.value), "Decrypted value not same as original")


@step("Encrypt some bytes")
@async_run_until_complete
async def step_impl(context: Context):
    assert_that(hasattr(context, "host"), equal_to(True), "Encrypter service host not set")
    assert_that(hasattr(context, "port"), equal_to(True), "Encrypter service port not set")
    assert_that(hasattr(context, "value"), equal_to(True), "String to encrypt not set")

    async with _Client(host=context.host, port=context.port) as client:
        encoded = b64encode(context.value.encode("utf-8"))
        value = await client.encrypt_bytes(encoded)
        assert_that(len(value), not_(equal_to(0)), "Encrypted value of zero length")
        context.encrypted = value


@step("Decrypt the encrypted string back to bytes")
@async_run_until_complete
async def step_impl(context):
    assert_that(hasattr(context, "host"), equal_to(True), "Encrypter service host not set")
    assert_that(hasattr(context, "port"), equal_to(True), "Encrypter service port not set")
    assert_that(hasattr(context, "value"), equal_to(True), "String to encrypt not set")

    async with _Client(host=context.host, port=context.port) as client:
        value = await client.decrypt_bytes(context.encrypted)
        assert_that(b64decode(value).decode("utf-8"), equal_to(context.value), "Decrypted value not same as original")
